c-api/compat-5.3.c

```diff
@@ -336,22 +392,155 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
     return 1;
   }
   else {
     lua_pushnil(L);
+#if defined(__GLIBC__) || defined(_POSIX_VERSION) || defined(__APPLE__) || (defined(__GNUC__) && (__GNUC__ < 6) && defined (__MINGW32__))
+    // use strerror_r here, because it's available on these specific platforms
+    char buf[512];
+    strerror_r(en, buf, 1024);
+    s = buf;
+#elif defined(_MSC_VER)
+    // for MSVC and others, use strerror_s since it's provided by default by the libraries
+    char buf[512];
+    strerror_s(buf, 1024, en);
+    s = buf;
+#else
+    s = strerror(en);
+#endif
     if (fname)
-      lua_pushfstring(L, "%s: %s", fname, strerror(en));
+      lua_pushfstring(L, "%s: %s", fname, s);
     else
-      lua_pushstring(L, strerror(en));
+      lua_pushstring(L, s);
     lua_pushnumber(L, (lua_Number)en);
     return 3;
   }
 }
 
+static const char* compat53_f_parser_handler (lua_State *L, void *data, size_t *size) {
+  struct compat53_f_parser_buffer *p = (struct compat53_f_parser_buffer*)data;
+  size_t readcount = fread(p->buffer + p->start, sizeof(*p->buffer), sizeof(p->buffer), p->file);
+  if (ferror(p->file) != 0) {
+    *size = 0;
+    return NULL;
+  }
+  p->start = 0;
+  *size = readcount;
+  return p->buffer;
+}
+
+COMPAT53_API int luaL_loadfilex(lua_State *L, const char *filename, const char *mode) {
+  size_t modesize;
+  FILE* fp;
+  errno_t err;
+  int c;
+  int returnvalue;
+  int allowbinary = 0;
+  int allowtext = 0;
+  if (mode == NULL) {
+    mode = "bt";
+    modesize = 2;
+  }
+  else {
+    modesize = strlen(mode);
+  }
+  if (modesize < 1) {
+    // mode is an empty string: boost it up
+    mode = "bt";
+    modesize = 2;
+  }
+  allowbinary = mode[0] == 'b'; 
+  allowtext = mode[0] == 't';
+  if (modesize > 1) {
+    allowbinary = mode[1] == 'b';
+    allowtext = mode[1] == 't';
+  }
+
+  if (!allowbinary || !allowtext) {
+    // invalid string: toss it out
+    return LUA_ERRFILE;
+  }
+  else if (allowbinary && allowtext) {
+     // both modes allowed: just pass it through
+     luaL_loadfile(L, filename);
+  }
+  // otherwise, we have to check manually ourselves
+#if defined(_MSC_VER)
+  err = fopen_s(&fp, filename, "rb");
+  if (err != 0) {
+    fclose(fp);
+    return LUA_ERRFILE;
+  }
+#else // fopen error on Visual C
+  fp = fopen(filename, "rb");
+  if (fp == NULL) {
+    fclose(fp);
+    return LUA_ERRFILE;
+  }
+#endif // fopen error on Visual C
+  c = fgetc(fp);
+  if (allowbinary) {
+    // if it's not a binary file, error it
+    if (c != '\x1b') {
+      fclose(fp);
+      return LUA_ERRSYNTAX; // bad syntax includes signature?
+    }
+  }
+  struct compat53_f_parser_buffer fbuf;
+  fbuf.file = fp;
+  fbuf.start = allowbinary != 0 ? 1 : 0;
+  returnvalue = lua_load(L, &compat53_f_parser_handler, &fbuf, filename);
+  fclose(fp);
+  return returnvalue;
+}
+
+COMPAT53_API int luaL_loadbufferx(lua_State *L, const char *buff, size_t sz, const char *name, const char *mode) {
+  size_t modesize;
+  FILE* fp;
+  errno_t err
```

Comment:
`errno_t` isn't valid in C.

Reply:
it is in c11, should still be `int` for c99 and prior.